### PR TITLE
Changed to only evaluate something as an Enum if we know that it is actually an Enum

### PR DESCRIPTION
**Pull request by jsight**

### CODE REVIEW

File: java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java [hunk 1]
```diff
@@ -326,69 +260,42 @@ private ClassReference processTypeBinding(ITypeBinding type, ResolutionStatus re
         return processTypeAsString(sourceString, packageName, className, resolutionStatus, referenceLocation, lineNumber, columnNumber, length, line);
     }
 
-    private ClassReference processTypeAsEnum(ITypeBinding type, ResolutionStatus resolutionStatus,
-                TypeReferenceLocation referenceLocation, int lineNumber, int columnNumber, int length, String line)
+    private ClassReference processTypeAsEnum(ITypeBinding typeBinding, Expression expression, ResolutionStatus resolutionStatus, int lineNumber, int columnNumber, int length, String line)
```
Comment (mareknovotny) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:263]: ok, so recognition from expression is much better

File: java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java [hunk 2]
```diff
@@ -625,35 +516,14 @@ public boolean visit(FieldDeclaration node)
...
-                        // the Enum Class and Constant doesn't have to be FQCN as package could miss 
-                        if (packageAndClassName.constantName != null) 
-                        {
-                            //additionally add Enum Constant type reference if found in String
-                            // verify it is ENUM and pass it ??? FIXME: is it really accesible the enum constant here or just ENUM
-                            processStringAsEnum(packageAndClassName, ResolutionStatus.RECOVERED, TypeReferenceLocation.ENUM_CONSTANT, lineNumber, columnNumber, node.getLength(), node.toString());
-                        }
-                    }
-                }
-            }
```
Comment (mareknovotny) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:475]: what if there is used statically imported enum constant? is that handled somewhere else?

File: java-ast/tests/src/test/java/org/jboss/windup/ast/java/test/JavaEnumerationScanningTest.java [hunk 3]
```diff
@@ -33,22 +35,25 @@ public void testEnumerationUsage()
...
+                ResolutionStatus.RESOLVED, TypeReferenceLocation.ENUM_CONSTANT, 31, 41, 21, "new EnumerationClassUsage(AccessMode.READ)");
+        Assert.assertTrue(references.contains(referenceTest));
     }
     
     @Test
     public void testEnumConstWithoutClassOnClasspath()
     {
-        List<ClassReference> references = ASTProcessor.analyze(getLibraryPaths(), getSourcePaths(),
+        Set<String> libraryPaths = new HashSet<>();
+        libraryPaths.add("src/test/resources/testclasses/enumeration/dependency/hibernate-search-engine-5.5.3.Final.jar");
```
Comment (mareknovotny) [inline java-ast/tests/src/test/java/org/jboss/windup/ast/java/test/JavaEnumerationScanningTest.java:54]: so do we not to try parse/analyze enums even they are not on classpath? This test is then wrongly named or am I wrong?

File: java-ast/tests/src/test/resources/testclasses/enumeration/EnumConstClassNotOnClasspath.java [hunk 4]
```diff
@@ -1,7 +1,7 @@
 package testclasses.enumeration;
 
 import org.hibernate.search.backend.configuration.impl.IndexWriterSetting;
-static import org.hibernate.search.backend.configuration.impl.IndexWriterSetting.MAX_THREAD_STATES;
+import static org.hibernate.search.backend.configuration.impl.IndexWriterSetting.MAX_THREAD_STATES;
```
Comment (mareknovotny) [inline java-ast/tests/src/test/resources/testclasses/enumeration/EnumConstClassNotOnClasspath.java:4]: bloody error :D.

File: java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java (hunk 2 above)
Comment (jsight) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:475]: JDT deals with that case automatically. The enum itself needs to be on the analysis path (or the additional associated library jars), though in order for jdt to know that it is an enum.

File: java-ast/tests/src/test/java/org/jboss/windup/ast/java/test/JavaEnumerationScanningTest.java (hunk 3 above)
Comment (jsight) [inline java-ast/tests/src/test/java/org/jboss/windup/ast/java/test/JavaEnumerationScanningTest.java:54]: Well, they have to be in JDT's analysis path, but not necessarily on the classpath. You couldn't Class.forName it in this case, but JDT will know about it.

The other test handles cases where it isn't on JDT's analysis path but is in the classloader.

File: java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java [hunk 5]
```diff
@@ -1131,17 +991,20 @@ public boolean visit(MethodInvocation node)
...
             resolutionStatus = ResolutionStatus.RESOLVED;
-            ITypeBinding[] arguments = resolveTypeBinding.getParameterTypes();
+            ITypeBinding[] argumentTypeBindings = resolveTypeBinding.getParameterTypes();
+            List<Expression> arguments = node.arguments();
 
-            for (ITypeBinding type : arguments)
+            int index = 0;
+            for (ITypeBinding type : argumentTypeBindings)
             {
+                Expression expression = arguments.get(index++);
```
Comment (mareknovotny) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:1000]: this is dangerous as it doesn't check the arguments size while iterating through argumentTypeBindings. This fails for some method invocations like https://docs.jboss.org/hibernate/search/4.3/api/org/hibernate/search/FullTextSession.html#createFullTextQuery(org.apache.lucene.search.Query,%20java.lang.Class...)
where are variable arguments.

Comment (jsight) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:1000]: Good point... I assume that you can fix this during the merge? Or should I update the PR?

Comment (mareknovotny) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:1000]: I already fixed it in my PR, I took your commit and did a few changes there too

Comment (jsight) [inline java-ast/addon/src/main/java/org/jboss/windup/ast/java/ReferenceResolvingVisitor.java:1000]: Oh, good... thanks! :)

**Pull request closed without merging**